core/src/main/java/org/apache/calcite/rel/core/Sort.java

```diff
@@ -157,6 +157,11 @@ public RelNode accept(RexShuttle shuttle) {
     return copy(traitSet, getInput(), collation, offset, fetch);
   }
 
+  @Override public boolean isEnforcer() {
+    return offset == null && fetch == null
```

Comment:
why offset and fetch have to be null? I feel that isEnforcer should be something passed down when the RelNode is created. It's hard to tell by itself that if an operator is served as an enforcer.

Reply:
If they are not null, that means it is a LIMIT operator, which is not an enforcer. I already give the definition of Enforcer in RelNode interface. An enforcer should be known when it is created, if the operator can't tell by itself it is an enforcer or not, it is either the design's problem, or just leave it as a non-enforcer.

Reply:
This is the definition from you - "As an enforcer, the operator must be created only when required traitSet is not satisfied by its input." So it sounds to me that only the caller (a RelOptRule or the framework) who creates the RelNode would know if the RelNode is a converter or not. If this information is not passed into the RelNode, how can we just derive this information from RelNode itself? In this particular case, if Sort is used in ORDER BY ... LIMIT ... scenario, it's still an enforcer. No?

Reply:
Good question, the answer is no. The sort in your example is created no matter it satisfies parent's required trait or not. So this is not an enforcer.

Reply:
What if just ORDER BY ...? So the Sort operator all the sudden become an enforcer when LIMIT is not presented?

Reply:
Yes, order by is an enforcer. Because the top required trait is something like EnumerableConvention with a collation.

Reply:
And ORDER BY ... LIMIT doesn't have such required trait?

Reply:
No

Reply:
I don't understand. In a lot of systems, the only difference is an extra limit operator on top of the sort. I don't get why it doesn't have required trait.

Reply:
Basically it is a limit operator. The parent of LIMIT doesn't require anything from it.

Reply:
The collation the required by the root, same as the ORDER BY case.

Reply:
No, the collation is required by LIMIT.

Reply:
I know it is confusing, but it is the design (Sort operator mixes both sort and limit operator) that leads to the confusion. We should have a separate LIMIT operator, SORT operator should not have limit and offset.

Reply:
When we have a query `select * from foo limit 5`, there is still a sort operator in the plan, but it doesn't do any sort work.